### PR TITLE
Using our subject/cache to fix the init race I introduced

### PR DESCRIPTION
**Pull request by jamesdaniels**
In my prior refactor I introduced two regressions:

1. `useFirestore()`, etc. would not throw a Promise if the init callback from `preloadFirestore` was still awaiting. This is important for `firestore.enablePersistence()` etc. to work correctly.
1. The init callback would not run if you called preload on a second app, since it was checking merely for existence on the namespace. We'll lean on the JS SDK's errors for double initialization.
1. It turns out that since we're side-effect free, webpack "might" copy our module, so our cache needs to be "global". I'm now attaching our preload cache to `globalThis`.

### CODE REVIEW

File: reactfire/firebaseApp/index.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ export * from './sdk';
 type FirebaseAppContextValue = firebase.app.App;
 
 // INVESTIGATE I don't like magic strings, can we have export this in js-sdk?
-export const DEFAULT_APP_NAME = '[DEFAULT]';
+const DEFAULT_APP_NAME = '[DEFAULT]';
```
Comment (jhuleatt) [inline reactfire/firebaseApp/index.tsx]: looks like this might not be needed any more?

File: sample/package.json [hunk 2]
```diff
@@ -5,6 +5,7 @@
   "dependencies": {
     "@types/jest": "^24.9.0",
     "firebase": "^7.9.1",
+    "js-levenshtein": "^1.1.6",
```
Comment (jhuleatt) [inline sample/package.json]: why is this needed?

**Comment by jhuleatt** [comment]
Want to take a shot at writing a firestore test to confirm that preloadFirestore safely sets the settings before any other calls are made?

File: reactfire/firebaseApp/index.tsx (hunk 1 above)
Comment (jamesdaniels) [reply reactfire/firebaseApp/index.tsx]: I'm still using within this module (`FirebaseAppProvider`) no longer using in `sdk.ts`, thus dropping the export.

File: sample/package.json (hunk 2 above)
Comment (jamesdaniels) [reply sample/package.json]: TBH, not sure, new peer dep somewhere in the tree. The demo app wouldn't compile without it.

**Comment by jamesdaniels** [comment]
@jhuleatt def. Going to write some defensive tests here to make sure we don't regress again. Also will permutate over default vs. contextual and multiple apps.

File: sample/package.json (hunk 2 above)
Comment (jamesdaniels) [reply sample/package.json]:
```
Failed to compile.

../node_modules/react-dev-utils/webpackHotDevClient.js
Error: [BABEL] /Users/james/Code/Firebase/reactfire/node_modules/react-dev-utils/webpackHotDevClient.js: Cannot find module 'js-levenshtein' (While processing: "/Users/james/Code/Firebase/reactfire/node_modules/babel-preset-react-app/dependencies.js")
```

File: reactfire/firebaseApp/index.tsx (hunk 1 above)
Comment (jhuleatt) [reply reactfire/firebaseApp/index.tsx]: 👍

File: sample/package.json (hunk 2 above)
Comment (jhuleatt) [reply sample/package.json]: Huh. I bumped the react version on master - maybe you need a fresh `yarn install`?
